x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java

```diff
@@ -188,39 +197,68 @@ void loadQueuedModels() {
             );
             if (unassignedIndices.size() > 0) {
                 logger.trace("not loading models as indices {} primary shards are unassigned", unassignedIndices);
+                onFinish.run();
                 return;
             }
         }
-        logger.trace("attempting to load all currently queued models");
-        // NOTE: As soon as this method exits, the timer for the scheduler starts ticking
-        Deque<TrainedModelDeploymentTask> loadingToRetry = new ArrayDeque<>();
-        while ((loadingTask = loadingModels.poll()) != null) {
-            final String deploymentId = loadingTask.getDeploymentId();
-            if (loadingTask.isStopped()) {
-                if (logger.isTraceEnabled()) {
-                    String reason = loadingTask.stoppedReason().orElse("_unknown_");
-                    logger.trace("[{}] attempted to load stopped task with reason [{}]", deploymentId, reason);
+
+        // deploy the models synchronously by flattening loadingModels into a chain of listeners
+        // if we need to retry, wait to add it back to loadingModels until after we finish this batch of deployments
+        // if someone calls stop halfway through, abandon this entire chain
+        var loadingToRetry = new ConcurrentLinkedDeque<TrainedModelDeploymentTask>();
+        var deploymentChain = SubscribableListener.<Void>newSucceeded(null);
+        while (loadingModels.isEmpty() == false) {
+            var loadingTask = loadingModels.poll();
+            deploymentChain = deploymentChain.andThen((l, r) -> {
+                if (stopped) {
+                    // don't bother calling the listener, the lifecycle will not resume the instance of this class
+                    return;
                 }
-                continue;
+
+                loadModel(loadingTask, l.delegateFailureAndWrap((ll, retry) -> {
+                    if (retry != null && retry) {
+                        loadingToRetry.offer(loadingTask);
+                    }
+                    ll.onResponse(null);
+                }));
+            });
+        }
+        deploymentChain.addListener(thenRun(() -> {
+            try {
+                loadingModels.addAll(loadingToRetry);
+            } finally {
+                onFinish.run();
             }
-            if (stopped) {
-                return;
+        }));
+    }
+
+    // using noop() here, because we want to ignore success or exceptions and always invoke the runnable
+    // runnable should not throw an exception though, it isn't caught anywhere
+    private static <T> ActionListener<T> thenRun(Runnable runnable) {
+        return ActionListener.runAfter(ActionListener.noop(), runnable);
```

Comment:
Why not `ActionListener.running()`?

Reply:
We should swap to it - initially I had this as `runBefore` so the runnable could safely throw an exception, then I thought failing silently might be confusing and swapped to `runAfter` without thinking that is effectively `ActionListener.running()`